Convert `Ascii` into an enum.

tormol opened the pull request:
Allows static initialization, nicer pattern matches and comparisons.

Naming conventions:

* alphabetic: just use it. breaks CamelCase convention, but worth it for simplicity.
* digits: prepend _ to make them valid identifiers, but still short.
* non-alphanumeric but visible: Use Wikipedia names CamelCased, but remove -Mark endings.
* rarely used control codes: use uppercase acronym to deter use.
* more commonly used control codes: Expand and CamelCase acronym to preserve meaning, eg LineFeed not NewLine or LineBreak.

I'm not certain which control codes should be expanded; Currently BEL and DEL are, but should they?

I'm not sure whether this is a breaking change.

tomprogrammer commented on src/ascii.rs:
```diff
@@ -52,13 +184,13 @@ impl Ascii {
     #[inline]
     pub fn as_byte(&self) -> u8 {
-        self.chr
+        unsafe{ transmute(*self) }
```
This conversion doesn't need an unsafe block because `*self as u8` works as expected.

tomprogrammer commented on src/ascii.rs:
```diff
@@ -23,7 +155,7 @@ impl Ascii {
     pub fn from(ch: char) -> Result<Ascii, ()> {
         if ch as u32 <= 0x7F {
-            return Ok( Ascii { chr: ch as u8 });
+            return Ok(unsafe{ ch.to_ascii_nocheck() });
```
Personally I would wrap the complete if-statement in the unsafe block. The if-clause is vital to the correctness of `ch.to_ascii_nocheck()`.

Do you mind changing it?

tomprogrammer commented:
I basically agree with the naming convention you chose, only minor discussion:

- In my opinion it's more "rustic" to write `Ascii::Nul` instead of `Ascii::NUL`. I propose changing the acronyms to only uppercase the first character.
- As you already mentioned it's difficult to categorize every control code by frequency of use. I guess that's not a severe question so if no discussion comes up I won't change the names.

The public interface isn't changed, in fact there is a new way to initialize `Ascii`. Therefore it isn't a breaking change. Nevertheless I will test that with crates depending on this crate, to be sure.

tormol commented:
Rebased on master to use wrapping_sub. Also added rustdocfor each enum variant in the first commit.

I decided to categorize control codes on whether they have an escape code listed in [this table](https://en.wikipedia.org/wiki/ASCII#ASCII_control_code_chart
)

While not rustic, I think uppercase names prevent confusion:
* is `At` `'@'` or some historical control code?
* is `Sub` some historical control code or `'-'`?

tomprogrammer commented:
I'm fine with your PR, thanks for your work!

I like the categorization based on the existence of an escape code, that makes it easier to guess how the variant is named. Your arguments also convinced me, that we should keep uppercase names for control codes. Indeed I've never seen control codes in CamelCase notation.

One minor point is that you used block comments `/* */` for the description of the variants. I'd like to change this to rustdoc comments `///`. Usually I look at the rustdoc documentation instead the plain code to find out how to use the library.

I can make that little change myself and thank your for your PR!

tomprogrammer commented:
Do you need the changes in this PR urgently? If so I will do a release, else I'll wait batching up some more changes I'm currently working on. ( #13 for example)

tormol commented:
No, I don't need the changes, and I'm working on another one.

`/**...*/` is also a doc comment, I used them here to avoid two lines for every variant.

tomprogrammer commented:
Oh, right, maybe I'll revert my changes that modify the doc-comments before the release.
